File: MPChartLib/src/main/java/com/github/mikephil/charting/renderer/RoundedBarChartRenderer.java
package com.github.mikephil.charting.renderer;

import android.graphics.Canvas;
import android.graphics.LinearGradient;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Shader;

import com.github.mikephil.charting.animation.ChartAnimator;
import com.github.mikephil.charting.buffer.BarBuffer;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.highlight.Range;
import com.github.mikephil.charting.interfaces.dataprovider.BarDataProvider;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.Utils;
import com.github.mikephil.charting.utils.ViewPortHandler;

/** @noinspection unused*/
public class RoundedBarChartRenderer extends BarChartRenderer {

	public RoundedBarChartRenderer(BarDataProvider chart, ChartAnimator animator, ViewPortHandler viewPortHandler) {
		super(chart, animator, viewPortHandler);
	}

	private final RectF mBarShadowRectBuffer = new RectF();
	private final float mRadius = 20f;
	private float roundedShadowRadius = 0f;
	private float roundedPositiveDataSetRadius = 0f;
	private float roundedNegativeDataSetRadius = 0f;

	public void setRoundedNegativeDataSetRadius(float roundedNegativeDataSet) {
		roundedNegativeDataSetRadius = roundedNegativeDataSet;
	}

	public void setRoundedShadowRadius(float roundedShadow) {
		roundedShadowRadius = roundedShadow;
	}

	public void setRoundedPositiveDataSetRadius(float roundedPositiveDataSet) {
		roundedPositiveDataSetRadius = roundedPositiveDataSet;
	}

	@Override
	protected void drawDataSet(Canvas c, IBarDataSet dataSet, int index) {
		initBuffers();
		Transformer trans = chart.getTransformer(dataSet.getAxisDependency());
		barBorderPaint.setColor(dataSet.getBarBorderColor());
		barBorderPaint.setStrokeWidth(Utils.convertDpToPixel(dataSet.getBarBorderWidth()));
		shadowPaint.setColor(dataSet.getBarShadowColor());
		boolean drawBorder = dataSet.getBarBorderWidth() > 0f;
		float phaseX = mAnimator.getPhaseX();
		float phaseY = mAnimator.getPhaseY();

		if (chart.isDrawBarShadowEnabled()) {
			shadowPaint.setColor(dataSet.getBarShadowColor());
			BarData barData = chart.getBarData();
			float barWidth = barData.getBarWidth();
			float barWidthHalf = barWidth / 2.0f;
			float x;
			int i = 0;
			double count = Math.min((double) (int) (double) ((float) dataSet.getEntryCount() * phaseX), dataSet.getEntryCount());
			while (i < count) {
				BarEntry e = dataSet.getEntryForIndex(i);
				x = e.getX();
				mBarShadowRectBuffer.left = x - barWidthHalf;
				mBarShadowRectBuffer.right = x + barWidthHalf;
				trans.rectValueToPixel(mBarShadowRectBuffer);
				if (!viewPortHandler.isInBoundsLeft(mBarShadowRectBuffer.right)) {
					i++;
					continue;
				}
                if (!viewPortHandler.isInBoundsRight(mBarShadowRectBuffer.left)) {
                    break;
                }
				mBarShadowRectBuffer.top = viewPortHandler.contentTop();
				mBarShadowRectBuffer.bottom = viewPortHandler.contentBottom();


				if (roundedShadowRadius > 0) {
					c.drawRoundRect(barRect, roundedShadowRadius, roundedShadowRadius, shadowPaint);
				} else {
					c.drawRect(mBarShadowRectBuffer, shadowPaint);
				}
				i++;
			}
		}

		BarBuffer buffer = barBuffers[index];
		buffer.setPhases(phaseX, phaseY);
		buffer.setDataSet(index);
		buffer.setInverted(chart.isInverted(dataSet.getAxisDependency()));
		buffer.setBarWidth(chart.getBarData().getBarWidth());
		buffer.feed(dataSet);
		trans.pointValuesToPixel(buffer.buffer);

		// if multiple colors has been assigned to Bar Chart
		if (dataSet.getColors().size() > 1) {

			for (int j = 0; j < buffer.size(); j += 4) {

                if (!viewPortHandler.isInBoundsLeft(buffer.buffer[j + 2])) {
                    continue;
                }

                if (!viewPortHandler.isInBoundsRight(buffer.buffer[j])) {
                    break;
                }

				if (chart.isDrawBarShadowEnabled()) {
                    if (roundedShadowRadius > 0) {
                        c.drawRoundRect(new RectF(buffer.buffer[j], viewPortHandler.contentTop(),
                                buffer.buffer[j + 2],
                                viewPortHandler.contentBottom()), roundedShadowRadius, roundedShadowRadius, shadowPaint);
                    } else {
                        c.drawRect(buffer.buffer[j], viewPortHandler.contentTop(),
                                buffer.buffer[j + 2],
                                viewPortHandler.contentBottom(), shadowPaint);
                    }
				}

				// Set the color for the currently drawn value. If the index
				mRenderPaint.setColor(dataSet.getColor(j / 4));

                if (roundedPositiveDataSetRadius > 0) {
                    c.drawRoundRect(new RectF(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
                            buffer.buffer[j + 3]), roundedPositiveDataSetRadius, roundedPositiveDataSetRadius, mRenderPaint);
                } else {
                    c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
                            buffer.buffer[j + 3], mRenderPaint);
                }
			}
		} else {

			mRenderPaint.setColor(dataSet.getColor());

			for (int j = 0; j < buffer.size(); j += 4) {

                if (!viewPortHandler.isInBoundsLeft(buffer.buffer[j + 2])) {
                    continue;
                }

                if (!viewPortHandler.isInBoundsRight(buffer.buffer[j])) {
                    break;
                }

				if (chart.isDrawBarShadowEnabled()) {
                    if (roundedShadowRadius > 0) {
                        c.drawRoundRect(new RectF(buffer.buffer[j], viewPortHandler.contentTop(),
                                buffer.buffer[j + 2],
                                viewPortHandler.contentBottom()), roundedShadowRadius, roundedShadowRadius, shadowPaint);
                    } else {
                        c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
                                buffer.buffer[j + 3], mRenderPaint);
                    }
				}

                if (roundedPositiveDataSetRadius > 0) {
                    c.drawRoundRect(new RectF(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
                            buffer.buffer[j + 3]), roundedPositiveDataSetRadius, roundedPositiveDataSetRadius, mRenderPaint);
                } else {
                    c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
                            buffer.buffer[j + 3], mRenderPaint);
                }
			}
		}


		boolean isSingleColor = dataSet.getColors().size() == 1;
		if (isSingleColor) {
			mRenderPaint.setColor(dataSet.getColor(index));
		}

		int j = 0;
		while (j < buffer.size()) {

			if (!viewPortHandler.isInBoundsLeft(buffer.buffer[j + 2])) {
				j += 4;
				continue;
			}

            if (!viewPortHandler.isInBoundsRight(buffer.buffer[j])) {
                break;
            }

			if (!isSingleColor) {
				mRenderPaint.setColor(dataSet.getColor(j / 4));
			}

			mRenderPaint.setShader(new LinearGradient(
					buffer.buffer[j],
					buffer.buffer[j + 3],
					buffer.buffer[j],
					buffer.buffer[j + 1],
					dataSet.getColor(j / 4),
					dataSet.getColor(j / 4),
					Shader.TileMode.MIRROR));

			mRenderPaint.setShader(new LinearGradient(
					buffer.buffer[j],
					buffer.buffer[j + 3],
					buffer.buffer[j],
					buffer.buffer[j + 1],
					dataSet.getColor(j / 4),
					dataSet.getColor(j / 4),
					Shader.TileMode.MIRROR));


			if ((dataSet.getEntryForIndex(j / 4).getY() < 0 && roundedNegativeDataSetRadius > 0)) {
				Path path2 = roundRect(new RectF(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
						buffer.buffer[j + 3]), roundedNegativeDataSetRadius, roundedNegativeDataSetRadius, true, true, true, true);
				c.drawPath(path2, mRenderPaint);
			} else if ((dataSet.getEntryForIndex(j / 4).getY() > 0 && roundedPositiveDataSetRadius > 0)) {
				Path path2 = roundRect(new RectF(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
						buffer.buffer[j + 3]), roundedPositiveDataSetRadius, roundedPositiveDataSetRadius, true, true, true, true);
				c.drawPath(path2, mRenderPaint);
			} else {
				c.drawRect(buffer.buffer[j], buffer.buffer[j + 1], buffer.buffer[j + 2],
						buffer.buffer[j + 3], mRenderPaint);
			}

			j += 4;
		}

	}

	@Override
	public void drawHighlighted(Canvas c, Highlight[] indices) {
		BarData barData = chart.getBarData();

		for (Highlight high : indices) {

			IBarDataSet set = barData.getDataSetByIndex(high.getDataSetIndex());

			if (set == null || !set.isHighlightEnabled()) {
				continue;
			}

			BarEntry e = set.getEntryForXValue(high.getX(), high.getY());

			if (!isInBoundsX(e, set)) {
				continue;
			}

			Transformer trans = chart.getTransformer(set.getAxisDependency());

			mHighlightPaint.setColor(set.getHighLightColor());
			mHighlightPaint.setAlpha(set.getHighLightAlpha());

			boolean isStack = high.getStackIndex() >= 0 && e.isStacked();

			final float y1;
			final float y2;

			if (isStack) {

				if (chart.isHighlightFullBarEnabled()) {

					y1 = e.getPositiveSum();
					y2 = -e.getNegativeSum();

				} else {

					Range range = e.getRanges()[high.getStackIndex()];

					y1 = range.from;
					y2 = range.to;
				}

			} else {
				y1 = e.getY();
				y2 = 0.f;
			}

			prepareBarHighlight(e.getX(), y1, y2, barData.getBarWidth() / 2f, trans);

			setHighlightDrawPos(high, barRect);

			Path path2 = roundRect(new RectF(barRect.left, barRect.top, barRect.right,
					barRect.bottom), mRadius, mRadius, true, true, true, true);

			c.drawPath(path2, mHighlightPaint);
		}
	}

	private Path roundRect(RectF rect, float rx, float ry, boolean tl, boolean tr, boolean br, boolean bl) {
		float top = rect.top;
		float left = rect.left;
		float right = rect.right;
		float bottom = rect.bottom;
		Path path = new Path();
        if (rx < 0) {
            rx = 0;
        }
        if (ry < 0) {
            ry = 0;
        }
		float width = right - left;
		float height = bottom - top;
        if (rx > width / 2) {
            rx = width / 2;
        }
        if (ry > height / 2) {
            ry = height / 2;
        }
		float widthMinusCorners = (width - (2 * rx));
		float heightMinusCorners = (height - (2 * ry));

		path.moveTo(right, top + ry);
        if (tr) {
            path.rQuadTo(0, -ry, -rx, -ry);//top-right corner
        } else {
            path.rLineTo(0, -ry);
            path.rLineTo(-rx, 0);
        }
		path.rLineTo(-widthMinusCorners, 0);
        if (tl) {
            path.rQuadTo(-rx, 0, -rx, ry); //top-left corner
        } else {
            path.rLineTo(-rx, 0);
            path.rLineTo(0, ry);
        }
		path.rLineTo(0, heightMinusCorners);

        if (bl) {
            path.rQuadTo(0, ry, rx, ry);//bottom-left corner
        } else {
            path.rLineTo(0, ry);
            path.rLineTo(rx, 0);
        }

		path.rLineTo(widthMinusCorners, 0);
		if (br)
			path.rQuadTo(rx, 0, rx, -ry); //bottom-right corner
		else {
			path.rLineTo(rx, 0);
			path.rLineTo(0, -ry);
		}

		path.rLineTo(0, -heightMinusCorners);
		path.close();
		return path;
	}
}

File: MPChartLib/src/main/java/com/github/mikephil/charting/renderer/RoundedBarChartRenderer.kt
package com.github.mikephil.charting.renderer

import android.graphics.Canvas
import android.graphics.LinearGradient
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Shader
import com.github.mikephil.charting.animation.ChartAnimator
import com.github.mikephil.charting.highlight.Highlight
import com.github.mikephil.charting.interfaces.dataprovider.BarDataProvider
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import com.github.mikephil.charting.renderer.BarChartRenderer
import com.github.mikephil.charting.utils.Utils
import com.github.mikephil.charting.utils.ViewPortHandler
import kotlin.math.abs
import kotlin.math.min

class RoundedBarChartRenderer(
    chart: BarDataProvider,
    animator: ChartAnimator?,
    viewPortHandler: ViewPortHandler?
) : BarChartRenderer(chart, animator, viewPortHandler) {

    private val shadowRect = RectF()
    private val tmpPts     = FloatArray(4)
    private val ovalPath   = Path()

    private val defaultRadius = 20f

    private var roundedShadowRadius          = 0f
    private var roundedPositiveDataSetRadius = 0f
    private var roundedNegativeDataSetRadius = 0f
    private var useAutoFullRadius            = false

    /** If true, corner radii = half the bar’s screen‐pixel width at draw‐time. */
    fun setUseAutoFullRadius(useAuto: Boolean) {
        useAutoFullRadius = useAuto
    }

    fun setRoundedShadowRadius(r: Float) {
        roundedShadowRadius = r
    }

    fun setRoundedPositiveDataSetRadius(r: Float) {
        roundedPositiveDataSetRadius = r
    }

    fun setRoundedNegativeDataSetRadius(r: Float) {
        roundedNegativeDataSetRadius = r
    }

    override fun drawDataSet(c: Canvas, dataSet: IBarDataSet, index: Int) {
        initBuffers()
        val trans    = chart.getTransformer(dataSet.axisDependency) ?: return
        val handler  = viewPortHandler

        val phaseX = animator.phaseX
        val phaseY = animator.phaseY

        // 1) auto‐radius?
        if (useAutoFullRadius) {
            val halfVal = chart.barData.barWidth / 2f
            tmpPts[0] = 0f; tmpPts[1] = 0f
            tmpPts[2] = halfVal; tmpPts[3] = 0f
            trans.pointValuesToPixel(tmpPts)
            val pxHalf = abs(tmpPts[2] - tmpPts[0])
            roundedShadowRadius          = pxHalf
            roundedPositiveDataSetRadius = pxHalf
            roundedNegativeDataSetRadius = pxHalf
        }

        // 2) prep paints
        barBorderPaint.color       = dataSet.barBorderColor
        barBorderPaint.strokeWidth = Utils.convertDpToPixel(dataSet.barBorderWidth)
        shadowPaint.color          = dataSet.barShadowColor

        // 3) draw shadows
        if (chart.isDrawBarShadowEnabled) {
            val barWidth = chart.barData.barWidth
            val half     = barWidth / 2f
            val count    = min((dataSet.entryCount * phaseX).toInt(), dataSet.entryCount)
            for (i in 0 until count) {
                dataSet.getEntryForIndex(i)?.let { e ->
                    val x = e.x
                    shadowRect.left  = x - half
                    shadowRect.right = x + half
                    trans.rectValueToPixel(shadowRect)

                    if (!handler.isInBoundsLeft(shadowRect.right) ||
                        !handler.isInBoundsRight(shadowRect.left)) return@let

                    shadowRect.top    = handler.contentTop()
                    shadowRect.bottom = handler.contentBottom()

                    if (roundedShadowRadius > 0f) {
                        c.drawRoundRect(shadowRect, roundedShadowRadius, roundedShadowRadius, shadowPaint)
                    } else {
                        c.drawRect(shadowRect, shadowPaint)
                    }
                }
            }
        }

        // 4) feed & transform
        val buffer = barBuffers[index] ?: return
        buffer.setPhases(phaseX, phaseY)
        buffer.setDataSet(index)
        buffer.setInverted(chart.isInverted(dataSet.axisDependency))
        buffer.setBarWidth(chart.barData.barWidth)
        buffer.feed(dataSet)
        trans.pointValuesToPixel(buffer.buffer)

        val singleColor = dataSet.colors.size == 1

        // 5a) multi‐color bars
        if (!singleColor) {
            var j = 0
            while (j < buffer.size()) {
                val left   = buffer.buffer[j]
                val top    = buffer.buffer[j + 1]
                val right  = buffer.buffer[j + 2]
                val bottom = buffer.buffer[j + 3]

                if (!handler.isInBoundsLeft(right)) { j += 4; continue }
                // if bar is off‐right, we're past visible, so stop
                if (!handler.isInBoundsRight(left)) break

                // shadow
                if (chart.isDrawBarShadowEnabled && roundedShadowRadius > 0f) {
                    ovalPath.reset()
                    ovalPath.addRoundRect(
                        RectF(left, handler.contentTop(), right, handler.contentBottom()),
                        roundedShadowRadius, roundedShadowRadius, Path.Direction.CW
                    )
                    c.drawPath(ovalPath, shadowPaint)
                }

                paintRender.color = dataSet.getColor(j / 4)
                if (roundedPositiveDataSetRadius > 0f) {
                    ovalPath.reset()
                    ovalPath.addRoundRect(
                        RectF(left, top, right, bottom),
                        roundedPositiveDataSetRadius,
                        roundedPositiveDataSetRadius,
                        Path.Direction.CW
                    )
                    c.drawPath(ovalPath, paintRender)
                } else {
                    c.drawRect(left, top, right, bottom, paintRender)
                }
                j += 4
            }
        }
        // 5b) single‐color bars
        else {
            paintRender.color = dataSet.color
            var j = 0
            while (j < buffer.size()) {
                val left   = buffer.buffer[j]
                val top    = buffer.buffer[j + 1]
                val right  = buffer.buffer[j + 2]
                val bottom = buffer.buffer[j + 3]

                if (!handler.isInBoundsLeft(right)) { j += 4; continue }
                if (!handler.isInBoundsRight(left)) break

                if (chart.isDrawBarShadowEnabled && roundedShadowRadius > 0f) {
                    ovalPath.reset()
                    ovalPath.addRoundRect(
                        RectF(left, handler.contentTop(), right, handler.contentBottom()),
                        roundedShadowRadius, roundedShadowRadius, Path.Direction.CW
                    )
                    c.drawPath(ovalPath, shadowPaint)
                }

                if (roundedPositiveDataSetRadius > 0f) {
                    ovalPath.reset()
                    ovalPath.addRoundRect(
                        RectF(left, top, right, bottom),
                        roundedPositiveDataSetRadius,
                        roundedPositiveDataSetRadius,
                        Path.Direction.CW
                    )
                    c.drawPath(ovalPath, paintRender)
                } else {
                    c.drawRect(left, top, right, bottom, paintRender)
                }
                j += 4
            }
        }

        // 6) gradient overlay
        var j = 0
        if (singleColor) paintRender.color = dataSet.getColor(index)
        while (j < buffer.size()) {
            val left   = buffer.buffer[j]
            val top    = buffer.buffer[j + 1]
            val right  = buffer.buffer[j + 2]
            val bottom = buffer.buffer[j + 3]

            if (!handler.isInBoundsLeft(right)) { j += 4; continue }
            if (!handler.isInBoundsRight(left)) break

            if (!singleColor) paintRender.color = dataSet.getColor(j / 4)
            paintRender.shader = LinearGradient(
                left, bottom, left, top,
                paintRender.color, paintRender.color,
                Shader.TileMode.MIRROR
            )

            val entryY = dataSet.getEntryForIndex(j / 4)?.y ?: 0f
            val radius = if (entryY < 0f) roundedNegativeDataSetRadius else roundedPositiveDataSetRadius
            if (radius > 0f) {
                ovalPath.reset()
                ovalPath.addRoundRect(
                    RectF(left, top, right, bottom),
                    radius, radius, Path.Direction.CW
                )
                c.drawPath(ovalPath, paintRender)
            } else {
                c.drawRect(left, top, right, bottom, paintRender)
            }
            j += 4
        }
        paintRender.shader = null
    }

    override fun drawHighlighted(c: Canvas, indices: Array<Highlight>) {
        // 1) early exits
        val handler = viewPortHandler
        val barData = chart.barData

        for (h in indices) {
            // 2) only highlight enabled sets
            val set = barData.getDataSetByIndex(h.dataSetIndex) ?: continue
            if (!set.isHighlightEnabled) continue

            // 3) find the matching Entry
            val e = set.getEntryForXValue(h.x, h.y) ?: continue
            if (!isInBoundsX(e, set)) continue

            // 4) compute the y‐range of the highlight (stack vs. normal)
            val isStack = h.stackIndex >= 0 && e.isStacked
            val (y1, y2) = if (isStack) {
                if (chart.isHighlightFullBarEnabled) {
                    e.positiveSum to -e.negativeSum
                } else {
                    val range = e.ranges[h.stackIndex]
                    range.from to range.to
                }
            } else {
                e.y to 0f
            }

            // 5) transform values to pixel‐rect
            val trans = chart.getTransformer(set.axisDependency) ?: continue
            prepareBarHighlight(
                e.x,
                y1,
                y2,
                barData.barWidth / 2f,
                trans
            )

            // 5b) record the center/top into the Highlight object so markers can be drawn
            setHighlightDrawPos(h, barRect)

            // 6) clip any highlights that are fully off‐screen
            if (!handler.isInBoundsLeft(barRect.right) ||
                !handler.isInBoundsRight(barRect.left) ||
                !handler.isInBoundsTop(barRect.bottom) ||
                !handler.isInBoundsBottom(barRect.top)
            ) {
                continue
            }

            // 7) choose corner radius
            val radius = if (useAutoFullRadius) {
                abs((barRect.right - barRect.left) / 2f)
            } else {
                defaultRadius
            }

            // 8) draw your rounded highlight
            paintHighlight.color = set.highLightColor
            paintHighlight.alpha = set.highLightAlpha
            c.drawRoundRect(barRect, radius, radius, paintHighlight)
        }
    }
}
